File: CSharpCodeBase/aiactions/ai_findparent.lua
local Class = require 'utils.hump.class'
local AINode = require 'utils.ailib.ainode'

-- We need to find parent, if parent not exists
local AI_FindParent = Class({__includes=AINode, init=function(self, entity)
  AINode.init(self, entity)
end})

function AI_FindParent:Visit()
  if self.status == NODE_READY then
    local parent = self.entity.relationship:GetInstance("parent")
    if parent == nil then
      parents = self.entity.vision:GetVisibleEntitiesByRSTag("parent")
      if #parents > 0 then
        self.entity.relationship:AddInstance("parent", parents[1], -1)
        self.status = NODE_SUCCESS
      else
        self.status = NODE_FAILURE
      end
    else
      self.status = NODE_SUCCESS
    end
  end
  self:Sleep(0.5)
end

return AI_FindParent

File: CSharpCodeBase/aiactions/ai_findparent.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class ai_findparent :ainode{
 // We need to find parent, if(parent not exists
 public void init(self, entity);
   AINode.init(self, entity);
 }})
 public void Visit(){
   if(this.status == NODE_READY  ){ 
     var parent = this.entity.relationship:GetInstance("parent");
     if(parent == null  ){ 
       parents = this.entity.vision:GetVisibleEntitiesByRSTag("parent");
       if(#parents > 0  ){ 
         this.entity.relationship:AddInstance("parent", parents[1], -1);
         this.status = NODE_SUCCESS;
       }else{
         this.status = NODE_FAILURE;
       }
     }else{
       this.status = NODE_SUCCESS;
     }
   }
   self:Sleep(0.5);
 }
 }}